xiaogpt/tts/volc.py

```diff
@@ -0,0 +1,141 @@
+logger = logging.getLogger(__name__)
+
+
+class VolcTTS(AudioFileTTS):
+    def __init__(self, mina_service, device_id, config):
+        super().__init__(mina_service, device_id, config)
+        self.token = get_token(config)
+        logger.info("Initializing VolcTTS {self.token}")
+    
+    async def make_audio_file(self, query: str, text: str) -> tuple[Path, float]:
+        tts_payload = json.dumps({
+            "text": text,
+            "speaker": self.config.volc_tts_speaker,
+            "audio_config": {
+                "format": "mp3",
+                "sample_rate": 24000,
+                "speech_rate": 0,
+            },
+        })
+        
+        
+        req = {
+            "appkey": self.config.volc_tts_app,
+            "token": self.token,
+            "namespace": "TTS",
+            "payload": tts_payload,
+        }
+        
+        resp = requests.post("https://sami.bytedance.com/api/v1/invoke", json=req)
+        try:
+            sami_resp = resp.json()
+            logger.info(f"sami_resp {resp.status_code}")
+            if resp.status_code != 200:
+                print(sami_resp)
+        except:
+            logger.error(f"Failed to get tts from volcengine with voice=zh {text}")
+        
+        if sami_resp["status_code"] == 20000000 and len(sami_resp["data"]) > 0:
+            audio_data = base64.b64decode(sami_resp["data"])
+            with tempfile.NamedTemporaryFile(
+                suffix=".mp3", mode="wb", delete=False, dir=self.dirname.name
+            ) as f:
+                f.write(audio_data)
+                
+        return Path(f.name), calculate_tts_elapse(text)
+
+## fetch token and save it to file
+def get_token(config):
+    token_file = "/tmp/volc_token.json"
+    if not os.path.exists(token_file):
+        token = request_token_data(config)
+    else:
+        with open(token_file, "r") as f:
+            token = json.load(f)
+            if token['expires_at'] < time.time():
+                token = request_token_data(config)
+                
+    if not os.path.exists(token_file):
+        with open(token_file, "w") as f:
+            json.dump(token, f)
+    return token['token']
+
+
+AUTH_VERSION = 'volc-auth-v1'
+
+def request_token_data(config):
+    sami_service = SAMIService()
+    sami_service.set_ak(config.volc_accesskey)
+    sami_service.set_sk(config.volc_secretkey)
+
+    req = {"appkey": config.volc_tts_app, "token_version": AUTH_VERSION, "expiration": 24 * 3600}
+    token = sami_service.common_json_handler("GetToken", req)
+    logger.info(f"Got token from volcengine {token}")
+    return token
+
+
+class SAMIService(Service):
+    _instance_lock = threading.Lock()
+
+    def __new__(cls, *args, **kwargs):
+        if not hasattr(SAMIService, "_instance"):
+            with SAMIService._instance_lock:
+                if not hasattr(SAMIService, "_instance"):
+                    SAMIService._instance = object.__new__(cls)
+        return SAMIService._instance
```

Comment:
That's a bit overkill, is it heavy operation to instantiate a `SAMIService`? if not you can create the instance at module level, rather than using a Singleton pattern

Reply:
I used the codes from the [official document](https://www.volcengine.com/docs/6489/71995#python)

Later I will try to reduce the usage of `volcengine` SDK. Maybe simple HTTP request is ok.